### PR TITLE
Tutorials

### PR DESCRIPTION
**Pull request by marios155**
Added C++ tutorials for basic operations (Basis generation, LLL Reduction and BKZ Reduction) with appropriate commentary. Created test file to test all of the above against pre-generated results from latticegen and fplll APIs. It would be great to know if my Makefile.am configuration in /tests directory is correct, as I do not know much about Makefiles.

### CODE REVIEW

File: tests/test_tutorials.cpp [hunk 1]
```diff
@@ -0,0 +1,280 @@
...
+   either version 2.1 of the License, or (at your option) any later version.
+
+   fplll is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+   GNU Lesser General Public License for more details.
+
+   You should have received a copy of the GNU Lesser General Public License
+   along with fplll. If not, see <http://www.gnu.org/licenses/>. */
+
```
Comment (malb) [inline tests/test_tutorials.cpp]: You should claim copyright.

File: tests/test_tutorials.cpp [hunk 2]
```diff
@@ -0,0 +1,280 @@
...
+using namespace fplll;
+
+/**
+   @brief Read T from `input_filename`.
+   @param X T (T is usually a ZZ_mat<ZT> or a vector<Z_NR<ZT>>
+   @param input_filename
+   @return zero if the file is correctly read, 1 otherwise.
+*/
+
+template <class T> int read_file(T &X, const char *input_filename) {
```
Comment (malb) [inline tests/test_tutorials.cpp]: There's a `read_file` in `test_utils.h`

File: tests/test_tutorials.cpp [hunk 3]
```diff
@@ -0,0 +1,280 @@
...
+   ZZ_mat<ZT> test_matrix;
+   int status = 1, input = 1, rows = lattice_base.get_rows(), columns = lattice_base.get_cols();
+   input = read_file(test_matrix, input_name);
+   if (input != 1)
+   {
+      for (int i = 0; i < rows; i++)
+      {
+         for (int j = 0; j < columns; j++)
+         {
+            status = test_matrix[i][j] != lattice_base[i][j];
```
Comment (malb) [inline tests/test_tutorials.cpp]: You should OR the output onto `status` instead of assigning.

**Comment by cr-marcstevens** [comment]
Travis failed on check-style, please install program `clang-format` and run `make check-style`.

**Comment by cr-marcstevens** [comment]
You mentioned `Makefile.am`, but I don't see any changes to any `Makefile.am`.

**Pull request closed without merging**